### PR TITLE
Restaurant views templates

### PR DESCRIPTION
**Pull request by KarthikrajanTP**
- Created List and Detail view for restaurants
- Created ListView for spotlight, bookmarked and visited restaurants
- Added image field in Dish model
- Created DishListView and template
- Created forms to handle reviews

### CODE REVIEW

File: restaurants/models.py [hunk 1]
```diff
@@ -69,6 +69,7 @@ class Dish(models.Model):
     name = models.CharField(max_length=100)
     price = models.DecimalField(max_digits=10, decimal_places=2)
     dish_type = models.CharField(max_length=20, choices=DISH_TYPE_CHOICES)
+    image = models.ImageField(upload_to='dish_photos/', null=True, blank=True)
```
Comment (shantanutomarr) [inline restaurants/models.py]: Create a separate PR to add this model field and the related migration. Model changes should be done in a separate commit.

File: templates/restaurant_detail.html [hunk 2]
```diff
@@ -1,30 +1,134 @@
...
 
+        <!-- Rating and Review Section -->
+        <div class="mt-8">
+            <h3 class="text-2xl font-semibold mb-4">Rate and Review</h3>
+            <form method="post" class="space-y-4">
+                {% csrf_token %}
+                {{ review_form.non_field_errors }}
+                <div>
+                    {{ review_form.rating.label_tag }}
+                    {{ review_form.rating|add_class:"block w-full mt-1 rounded-lg border-gray-300 shadow-sm focus:border-blue-500 focus:ring focus:ring-blue-200 focus:ring-opacity-50" }}
```
Comment (shantanutomarr) [inline templates/restaurant_detail.html]: This is a great opportunity for you to learn about widget_tweaks package. We have used this package extensively in testpress codebase.
https://pypi.org/project/django-widget-tweaks/

Check ``{% render_field form.title class+="css_class_1 css_class_2" %}

File: restaurants/views.py [hunk 3]
```diff
@@ -12,7 +18,19 @@ class RestaurantListView(ListView):
...
+        if sort_option == 'cost_asc':
+            queryset = queryset.order_by('cost_for_two')
+        elif sort_option == 'cost_desc':
+            queryset = queryset.order_by('-cost_for_two')
+        elif sort_option == 'rating_asc':
+            queryset = queryset.order_by('rating')
+        elif sort_option == 'rating_desc':
+            queryset = queryset.order_by('-rating')
+
+        return queryset
```
Comment (shantanutomarr) [inline restaurants/views.py]: See if you can implement the django_filters for sorting too. I am attaching a screenshot of a filter used in testpress. Check the `o`  field Ignore everything else.

Screenshot from 2024-08-08 12-06-48

File: restaurants/views.py [hunk 4]
```diff
@@ -29,7 +47,96 @@ def get_context_data(self, **kwargs):
...
+        })
+
+    def post(self, request, pk, *args, **kwargs):
+        restaurant = get_object_or_404(Restaurant, pk=pk)
+        user_review = restaurant.reviews.filter(user=request.user).first()
+
+        if user_review:
+            form = ReviewForm(request.POST, instance=user_review)
+        else:
+            form = ReviewForm(request.POST)
```
Comment (shantanutomarr) [inline restaurants/views.py]: Extract this to a new method called something like `_is_update_review_form()`

See a better solution below. If instance is None the form will automatically create a new review.

File: restaurants/views.py [hunk 5]
```diff
@@ -29,7 +47,96 @@ def get_context_data(self, **kwargs):
...
+                review.bookmarked = True
+            else:
+                review.bookmarked = False
+            
+            if 'visited' in request.POST:
+                review.visited = True
+            else:
+                review.visited = False
+
+            review.save()
```
Comment (shantanutomarr) [inline restaurants/views.py]: Extract these to separate method having intent revealing name.

File: restaurants/views.py [hunk 6]
```diff
@@ -29,7 +47,96 @@ def get_context_data(self, **kwargs):
...
+
+class DishListView(ListView):
+    model = Dish
+    template_name = 'dish_list.html'
+    context_object_name = 'dishes'
+    paginate_by = 9
+
+    def get_queryset(self):
+        restaurant_id = self.kwargs.get('pk')
+        return Dish.objects.filter(restaurant_id=restaurant_id)
```
Comment (shantanutomarr) [inline restaurants/views.py]:
```suggestion
        return Dish.objects.filter(restaurant_id=self.kwargs.get('pk'))
```

File: restaurants/views.py [hunk 7]
```diff
@@ -29,7 +47,96 @@ def get_context_data(self, **kwargs):
...
-    template_name = 'restaurant_detail.html'
-    context_object_name = 'restaurant'
\ No newline at end of file
+    context_object_name = 'restaurants'
+    template_name = 'restaurant_common_list.html'
+    page_title = 'Visited Restaurants'
+
+    def get_queryset(self):
+        user = self.request.user
+        return Restaurant.objects.filter(reviews__user=user, reviews__visited=True).distinct()
```
Comment (shantanutomarr) [inline restaurants/views.py]:
```suggestion
        return Restaurant.objects.filter(reviews__user=self.request.user, reviews__visited=True).distinct()
```

**Review by shantanutomarr — CHANGES_REQUESTED** [review]

File: restaurants/views.py (hunk 4 above)
Comment (shantanutomarr) [inline restaurants/views.py]:
```suggestion
        user_review = restaurant.reviews.filter(user=request.user).first()
        form = ReviewForm(request.POST, instance=user_review)
      
```

**Comment by KarthikrajanTP** [comment]
Closing this PR to see the changes done by merging the models branch to master branch.

**Review by shantanutomarr — APPROVED** [review]

**Pull request merged**